New `experimental_shell_command`

kaos opened the pull request:
The new `experimental_shell_command` added to the `shell` backend allows running arbitrary commands during `pants` execution.

This target is for introducing side effects in the build, either as new or modified files, calling out to external services or managing some other state. It remains important to ensure idempotency however, as the command may be cancelled or retried on the sole discretion of Pants.

For those familiar with Bazel, the `experimental_shell_command` has similarities with the [Bazel `genrule`](https://docs.bazel.build/versions/main/be/general.html#genrule).

Fixes #3734

Example BUILD file usage:

```python
shell_library(name="build-tools")
experimental_shell_command(
  command="./build-util.sh -o output do-things",
  tools=["bash", "env", "cat", "curl", "tar"],
  outputs=["output/"],
  dependencies=[":build-tools"],
)

```

The `dependencies` will pull in scripts from `shell_library`, arbitrary files from `files` and other `experimental_shell_command` targets, the `outputs` lists directories and files to capture, which may be included by consuming targets, and `tools` lists all required executables that `command` may be using.

The [`[shell-setup].executable-search-paths`](https://www.pantsbuild.org/v2.8/docs/reference-shell-setup#section-executable-search-paths) option is used when finding the specified `tools`.

kaos commented:
@stuhood any thoughts on the direction of this?

stuhood commented on src/python/pants/core/util_rules/gen_rule.py:
```diff
@@ -0,0 +1,58 @@
+
+@rule(desc="Running gen_rule", level=LogLevel.DEBUG)
+async def run_gen_rule(gen_rule: GenRuleTarget) -> ProcessResult:
+    args = shlex.split(cast(str, gen_rule[GenRuleCommandField].value))
+    binary_request = BinaryPathRequest(
+        binary_name=args[0],
+        search_path=SearchPath(("/usr/bin", "/bin", "/usr/local/bin")),
+    )
```
Interesting! I didn't expect this to open with tool discovery, but I suppose that that might make sense?

One thing that occurs to me though is that it is fairly likely to need multiple tools available to a shell script (for example, if arg0 is `bash`, then you are likely to also need `grep`, `cat` etc). Perhaps rather than automatically discovering `args[0]`, the target type should instead accept a list of tool names to resolve, and then expose them as environment variables that the command can use?

stuhood commented on src/python/pants/core/util_rules/gen_rule_test.py:
```diff
@@ -0,0 +1,61 @@
+                """\
+                gen_rule(
+                  name="hello",
+                  sources=["message.txt"],
+                  command="cat message.txt",  # support $(sources) etc similar to what bazel does?
```
I'm slightly surprised that `gen_rule` has a `sources` field and no `deps` field... perhaps that's because bazel doesn't have a generic `files` target the way that pants does? But given that we have a generic `files` target, I wonder if this target should consume `dependencies` rather than having sources of its own?

For one thing, assuming that this target is implemented as [codegen](https://www.pantsbuild.org/docs/plugins-codegen) that generates `Files`, having a dependencies field would allow for chaining multiple `gen_rules` together via `dependencies`.

As to your question about a `sources` env var: yea, I think that that makes sense. It allows for usage of globs, for example.

kaos replied:
The lack of dependencies field is simply by omission to keep the size down. Replacing sources for dependencies makes sense. As you can always depend on a files target to get loose files in.

kaos replied:
Not entirely sure I get the connection how to support glob usage with a sources env var..?

stuhood replied:
> Not entirely sure I get the connection how to support glob usage with a sources env var..?

If you adjust the example in the PR to use globs, you'll see an issue:
```
sources=["*.txt"],
command="cat *.txt",
```
... won't work for a few reasons:
1. glob expansion usually happens in a shell, and `cat` won't have one here. So it will fail to find a literal file named `*.txt`
2. even if the process being invoked is `bash`, bash's glob expansion will not be equivalent to our own (which tries to match `git`'s syntax).

Having said that! I just realized that there is a bit of a dilemma here. Directly invoking `cat` still wouldn't work with an env var containing a list of sources... because env vars are expanded by shells. All of this might be why Bazel went with requiring that the top level process be `bash`? I could imagine having our own sort of templating/placeholder syntax other than env vars though, maybe...?

kaos replied:
Mmm, yes, Thatcher could be feasible. I'll expand on the example first, and try to introduce some glob use see how it pans out.

kaos replied:
I've gone with running the command with `bash -c`. This is the same concept used in Dockerfiles and their `RUN` instruction.

stuhood commented on src/python/pants/core/target_types.py:
```diff
@@ -324,6 +324,26 @@ async def package_archive_target(field_set: ArchiveFieldSet) -> BuiltPackage:
+        GenRuleCommandField,
+        Sources,
+    )
+    help = "Execute any external tool for its side effects."
```
Should probably specify that since the process is sandboxed, it can only really have external side effects. Also, as with the discussion on https://github.com/pantsbuild/pants/pull/12805#discussion_r705663991, it's worth calling out that its best for any external side effects (poking the network, etc) to be idempotent, since they may be retried or canceled.

stuhood commented on src/python/pants/core/target_types.py:
```diff
@@ -324,6 +324,26 @@ async def package_archive_target(field_set: ArchiveFieldSet) -> BuiltPackage:
+
+
+class GenRuleTarget(Target):
+    alias = "gen_rule"
```
Although we should definitely call out the connection to `gen_rule` in the docs, we should probably also choose a more useful name! I've never really know what the `gen` stands for ("generate"? "generic"?).

kaos replied:
How about `ext_rule`. `ext` being as diffuse as `gen`, could stand for "external", "extended" etc :P 

No, but more seriously, perhaps it could be called simply `make`. As a target for "making" stuff.. not to be confused with the tool named "make", though, so perhaps not that great any way.. oh, dang. Naming stuff is hard.

Eric-Arellano replied:
I think I'm still confused what the goal does. You run arbitrary bash commands and the generated files can be dependended upon by other targets? You can also use it to trigger side effects like starting a server?

kaos replied:
Yeah, as a way to process files midway through a rule graph. Haven't thought of starting servers or such. Could possibly be used for that, but not sure it is good practice (unclear life cycle management).

kaos replied:
```python
make(
  name="Mangle my files to proper format",
  dependencies=[":some-inputs", ":util-scripts"],
  output_files=["result.out"],
  tools=["awk", "jq", "sed"],
  command="./format-util *.in > result.out"
)
```

kaos replied:
Or, just call it `bash`, or `shell` even, as I'm leaning towards always executing the `command` using `bash -c <command>`, and if we want to open up for making what shell to use optional, then a more generic target name like `shell` makes sense to me.

```python
shell(
    ...,
    command="echo 'foo bar' > output",
    outputs=["output", "dest/"],
)
```

Eric-Arellano replied:
The plan is to rename `shell_library` to `shell_sources` and `shell_source`. So here, we could maybe have `shell_command`?

>  It seems like taking advantage of that symmetry would require that this target type is able to usefully depend on shell_library targets to source / import them...

I think it makes sense that dependencies should be populated in the context, such that you could indeed set your command to do something like `command=build-support/my_script.sh`.

stuhood replied:
I was hesitant about requiring this at first, but thinking about it more, it definitely aligns with best practice: your actual binary definition should be small, because it's much harder to test a binary than it is to test a library.

stuhood reviewed (commented):
Thanks a lot! This looks like the right track. Allowing `gen_rule` outputs to be consumed by other targets as `Files` is probably the most critical bit to implement.

stuhood commented:
Also curious whether @Eric-Arellano thinks this might need/want to integrate with [shell_library](https://www.pantsbuild.org/docs/reference-shell_library).

kaos commented on src/python/pants/core/target_types.py:
```diff
@@ -353,6 +354,58 @@ async def package_archive_target(field_set: ArchiveFieldSet) -> BuiltPackage:
+
+
+class GenRuleSources(Sources):
+    alias = "sources"
```
I think it makes sense to have a sources field, as a way to put an overly complicated command into a script file instead of having it inline in the BUILD file is a big win. Granted, it could be housed in a `files` target, but it feels a bit redundant to have an entire separate goal, for just one script file.

And seeing as the source field was required to be present any way, in order for code gen to work (see relocate files target) might as well use it.

See tests, for example use of the sources field.

Eric-Arellano replied:
Probably better in a `shell_library` (soon to be `shell_source`/`shell_sources`) target, then loaded with `dependencies`. I think it makes sense to get rid of `Sources` for this or make it private.

We don't want two ways of "owning" a .sh file imo.

kaos replied:
I have no strong preference either way. 

IIUC, I can't get rid of the `sources` field, in order to have codegen work. Is using a `_` prefix on the alias for the sources field marking it private?

Eric-Arellano replied:
Exactly

kaos commented on src/python/pants/core/util_rules/gen_rule.py:
```diff
@@ -0,0 +1,145 @@
+        ),
+    )
+
+    output = await Get(Snapshot, AddPrefix(result.output_digest, working_directory))
```
I'm not entirely sure how to treat the paths here... should we look at `uses_source_roots` or so, for instance..?

kaos replied:
This was a fix, to get the input files in a determined place, as without, they all ended up in the root, but I set the working dir to spec_path of the gen_rule target. That makes it hard to reason about where files end up.

stuhood replied:
This seems fine for now, yea.

Relatedly though: I think that we should temporarily put `experimental_` in the name of this target (`experimental_shell`, etc) to ease adjusting these semantics after folks gain some experience using it? We can deprecate the name and change the semantics in the future to stabilize it.

kaos commented on src/python/pants/core/util_rules/gen_rule_test.py:
```diff
@@ -0,0 +1,154 @@
+                    "mkdir",
+                    "tee",
+                  ],
+                  outputs=["message.txt", "res/"],
```
I prefer fewer fields but with a convention that all directories must end with `/`.

kaos commented on src/python/pants/core/util_rules/gen_rule_test.py:
```diff
@@ -0,0 +1,154 @@
+            "src/intro.txt": MSG[0],
+            "src/outro.txt": MSG[1],
+            "src/script": (
+                "$mkdir res && $cat *.txt > message.txt && $cat message.txt | $tee res/log.txt"
```
In order to run shellfmt etc on the command script, it's preferable to be able to put it in a dedicated file.

kaos commented on src/python/pants/core/util_rules/gen_rule_test.py:
```diff
@@ -0,0 +1,154 @@
+                  name="msg",
+                  tools=["cp", "echo"],
+                  outputs=["msg"],
+                  command="$cp ../a/msg . ; $echo 'gen_rule:b' >> msg",
```
I had a bit of struggle to get this right, as I was confused about where msg a's input file landed.

stuhood replied:
Yea, I think that finding the right semantics for this will definitely take some iteration. It looks like you've really done a great job experimenting here (thank you!), but probably safest to give the target an `experimental_` name until we've gotten more feedback.

kaos reviewed (commented):
I'd be happy to rename gen_rule => shell or something like that. Apart from that, I feel this is getting pretty close to done.

I don't think we have to consider the difference in glob behaviour between git style and shell style too much. As long as it is clear that the `command` line is parsed by a shell, it shouldn't be any surprises there.

We could potentially however provide more data in the environment, regarding inputs and outputs etc.

kaos commented on src/python/pants/core/util_rules/gen_rule.py:
```diff
@@ -0,0 +1,145 @@
+    result = await Get(
+        ProcessResult,
+        Process(
+            argv=(bash.path, "-c", command),
```
We always run all commands using a default shell, such as bash in this case, to get glob and env var expansion etc.

stuhood commented on src/python/pants/core/util_rules/gen_rule.py:
```diff
@@ -0,0 +1,145 @@
+    outputs = gen_rule[GenRuleOutputsField].value
+    source_globs = gen_rule[GenRuleSources].value or ()
+
+    if not (command and tools and outputs):
```
The docstring says that the command runs "for side effects": the purest definition of a "side effect" in my mind is an effect which is not an output, and instead mutates the world in some way. If that's really what you meant by side effect, then allowing commands without `outputs` to run would make sense.

But also, it seems like raising a `ValueError` for a noop target would make sense. You can do it directly on the `Target` definition similarly to #12925.

kaos replied:
Yeah, not output should def be allowed. There was an issue with running the Process though, that caught me a bit off guard, that had to do with digests (input or output), but don't recall the details right now. Will find out again if it was the output one..

stuhood commented on src/python/pants/core/util_rules/gen_rule.py:
```diff
@@ -0,0 +1,145 @@
+    own_sources, dep_sources = await MultiGet(
+        Get(
+            Digest,
+            PathGlobs(
+                [os.path.join(working_directory, glob) for glob in source_globs],
+                glob_match_error_behavior=GlobMatchErrorBehavior.error,
+                description_of_origin=f"{gen_rule.address}'s `{GenRuleSources.alias}` field",
+            ),
```
You can call [Sources.path_globs](https://github.com/pantsbuild/pants/blob/9c12fc656813d8938342b115cc8c445418c6d93e/src/python/pants/engine/target.py#L1431-L1451) for this.

kaos replied:
Ah, thanks :)

stuhood reviewed (approved):
This looks really great: thanks a lot!

I think that this is ready to land with an `experimental_` prefixed name.

Eric-Arellano commented:
> Also curious whether @Eric-Arellano thinks this might need/want to integrate with shell_library.

What might that look like? I think it makes sense to be separate. Fwict, `shell_library` describes shell source code checked in to disk. Iiuc, this new target does not require anything on-disk necessarily.

Eric-Arellano commented on src/python/pants/core/target_types.py:
```diff
@@ -353,6 +354,58 @@ async def package_archive_target(field_set: ArchiveFieldSet) -> BuiltPackage:
+    required = True
+    help = (
+        "Specify required executable tools that is being used.\n\n"
+        "Each tool will be available in the environment, e.g. the tool 'tar' should be used "
```
Do you mean this?

```suggestion
        "Each tool will be available in the environment via environment variables, e.g. the tool 'tar' should be used "
```

Eric-Arellano commented on src/python/pants/core/target_types.py:
```diff
@@ -353,6 +354,58 @@ async def package_archive_target(field_set: ArchiveFieldSet) -> BuiltPackage:
+    required = True
+    help = (
+        "Specify the shell command output files and directories.\n\n"
+        "Use a trailing slash on directory names."
```
```suggestion
        "Use a trailing slash on directory names, i.e. `my_dir/`."
```

Eric-Arellano commented on src/python/pants/core/target_types.py:
```diff
@@ -353,6 +354,58 @@ async def package_archive_target(field_set: ArchiveFieldSet) -> BuiltPackage:
+    )
+    help = (
+        "Execute any external tool for its side effects.\n"
+        "This may be retried and/or cancelled, so ensure it is idempotent."
```
I think this won't be intuitive for a lot of users. It'd be helpful to put some examples in here, including possibly the target definition.

Eric-Arellano commented on src/python/pants/core/util_rules/gen_rule.py:
```diff
@@ -0,0 +1,145 @@
+
+class GenerateFilesFromGenRuleRequest(GenerateSourcesRequest):
+    input = GenRuleSources
+    output = FilesSources
```
Note that `Files` are not (currently) included when building a `pex_binary` and `python_awslambda`. They're mostly helpful for tests and for `archive`. Is that okay?

kaos replied:
Ah, I'd like the output to be as generic as possible, is plains `Sources` better, then? 
Or rather, I want to make sure that whatever is produced by the `shell_command` (if we call it that) may be used as sources by another target.

Not sure if simply depending on a `shell_command`, from say a `python_library` is enough, or if you should include the output files from `shell_command` in the `python_library` sources field? But that may complain that the "glob doesn't match" any files, until the `shell_command` has been executed, if that is an issue?

Eric-Arellano reviewed (commented):
Cool! I looked mostly at the modeling of the target, not the actual rule implementation

kaos commented:
Seing now this as a (`experimental_`)`shell_command`, maybe it ought to live in the `shell` backend?

kaos commented:
I just realize that I really doesn't like to use `$tool` instead of just `tool`. That puts a lot of constraints on the script to run, and also complicates if you want to run it outside of a pants execution.

Just take the leading shebang, for example (is what prompted me down this lane to begin with):
```
#!/usr/bin/env bash
...
```
This won't work, but should instead be just
```
#!$bash
...
```

Ugh.

So, i propose that we setup a `.tools/bin/` folder, and symlink all requested tools in there, and list that in the search path `PATH=/path/to/sandbox/.tools/bin` so that ordinary `env` works, etc.

kaos commented:
- I moved the whole implementation of `experimental_shell_command` to the shell backend.
- Added a new field `log_output`, anticipating that it will be good to be able to print something informative from the command, in case it is only side-effecting without any `outputs` captured, at least.
- Sets up a `.bin` dir with symlinks to all tools specified for use, and points `PATH` at that .bin dir, so we can use `cd`, `cat` etc without having to use parameter expansion from the environment on them.

stuhood commented on src/python/pants/backend/shell/shell_command.py:
```diff
@@ -0,0 +1,156 @@
+    )
+
+    command_env = {
+        "TOOLS": " ".join(tools),
```
It's unlikely, but tools might have spaces in their paths. Should `shlex.escape` them before joining them.

stuhood reviewed (approved):
This looks really fantastic: thanks a lot @kaos!

Eric-Arellano commented on src/python/pants/core/register.py:
```diff
@@ -33,22 +33,22 @@ def rules():
         *package.rules(),
```
Thanks for fixing the ordering :)

kaos replied:
Yeah, it was hurting my eyes, and `C-space C-s ] C-p M-x so-li enter` is a mere 2.5 second fix :P

Eric-Arellano commented on src/python/pants/backend/shell/target_types.py:
```diff
@@ -192,6 +195,77 @@ async def generate_targets_from_shell_library(
+    help = dedent(
+        """\
+        Execute any external tool for its side effects.
+        This may be retried and/or cancelled, so ensure that it is idempotent.
```
Nit, should probably move below the example. This is a nuanced detail that doesn't explain _what_ this target is, only a limitation when using it. The example is really helpful to figure out _what_ it is.

Might also be worth adding a paragraph below the example explaining how you can depend on `shell_library` targets to run its script, or you can directly invoke Bash commands like `touch my_file.ext`.

Eric-Arellano replied:
Oh! Another key detail missing from this help, you must add this target to the `dependencies` of each consumer, such as your `python_tests`. When relevant, Pants will run your `command` and insert the `output_files` into that build's context.

kaos replied:
> Might also be worth adding a paragraph below the example explaining how you can depend on shell_library targets to run its script

Oh, wait, I don't think I fully understand what you mean here.

kaos replied:
Oh, could we get around this limitation, at least to some extent, if we implement the `GenerateTargetsRequest` union for shell_command as well (and generate file targets for any output files)?

stuhood replied:
IMO, only being able to bring in all of the output files of the script is totally fine for now. And `files` is almost certainly the right output type.

`GenerateTargetsRequest` wouldn't be able to fully determine which files existed without actually running the script, because `directory/` entries include the children of the directory.

Eric-Arellano replied:
> Oh, wait, I don't think I fully understand what you mean here.

What I mean is that if you set your command to be `./my_script.sh`, you should add a `shell_library` target which includes `my_script.sh` in its `sources` and include that in the `dependencies` field.

Eric-Arellano commented on src/python/pants/backend/shell/shell_command.py:
```diff
@@ -0,0 +1,156 @@
+
+class GenerateFilesFromShellCommandRequest(GenerateSourcesRequest):
+    input = ShellCommandSources
+    output = FilesSources
```
To respond to the earlier thread: I think this is the right choice for now to start, rather than `Sources`.

Note also that callers must set `enable_codegen=True` for this codegen to happen.

Eric-Arellano reviewed (approved):
This is awesome, great work!

stuhood commented:
With a fresh description and title, I think this is probably ready to land! Thanks again.

Eric-Arellano commented on src/python/pants/backend/shell/target_types.py:
```diff
@@ -247,22 +247,28 @@ class ShellCommand(Target):
-              outputs=["results/", "logs/my-script.log"],
-            )
+    help = (
+        "Execute any external tool for its side effects.\n"
+        + dedent(
```
Yay thanks for doing that formatting :) This is great because Pants wraps lines based on your terminal width, but we need to use implicit string concatenation for that to work properly.

(The docs also work better when using implicit string concatenation so that the browser can wrap for you.)

kaos replied:
Yes, I remember you told me this once before ;)

kaos commented:
Right, I forgot there is one more thing I'd like to include here before I feel it is ready to go.. that is I'd like to use the [executable-search-paths](https://www.pantsbuild.org/v2.8/docs/reference-shell-setup#section-executable-search-paths) option from the `shell-setup` subsystem, rather than using a hard coded list of the most common search paths.

Eric-Arellano commented on src/python/pants/backend/shell/target_types.py:
```diff
@@ -226,7 +226,9 @@ class ShellCommandToolsField(StringSequenceField):
         "Specify required executable tools that might be used.\n\n"
-        "Only the tools explicitly provided will be available on the search PATH."
+        "Only the tools explicitly provided will be available on the search PATH, "
+        "and these tools must be found on the paths provided by "
+        "[shell-setup].executable-search-paths (which defaults to the system PATH)."
```
Love it. Great detail!

benjyw replied:
I am so excited for this!

Random nit: Don't option names in config have to use underscores? i.e., `executable_search_paths`? We let them be with dashes on the cmd line (`--shell-setup-executable-search-paths`) because `--shell-setup-executable_search_paths` looks awful, but I think we at least encourage underscores in option names (but dashes in scope names, so it's still `shell-setup`).

Eric-Arellano replied:
Oh yeah true, I think it needs to be underscores for the option part

kaos commented:
I copied the PR description in as merge commit message, kept author info. I assume you can disable, and fixup if that is not what you want? :)

Eric-Arellano commented:
> I copied the PR description in as merge commit message, kept author info. I assume you can disable, and fixup if that is not what you want? :)

That sounds good! The most important part is the PR title being useful for the changelog, which this is. (Nit that our changelog is Markdown, so escaping with ` is useful)

For the description, copying in the PR description is great too. And totally fine to include author info. Also you can copy the `[ci skip-rust]` part when it's safe to.
